File: src/main/java/social/alone/server/event/repository/EventRepositoryImpl.java
package social.alone.server.event.repository;

import com.querydsl.core.types.Predicate;
import com.querydsl.jpa.JPQLQuery;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.support.QuerydslRepositorySupport;
import social.alone.server.event.Event;
import social.alone.server.event.QEvent;
import social.alone.server.event.type.Coordinate;
import social.alone.server.event.type.EventQueryParams;
import social.alone.server.user.User;

import java.util.List;
import java.util.Optional;

public class EventRepositoryImpl extends QuerydslRepositorySupport implements EventRepositoryCustom {

  public EventRepositoryImpl() {
    super(Event.class);
  }

  @Override
  public Page<Event> search(Pageable pageable, User user, EventQueryParams eventQueryParams) {
    QEvent event = QEvent.event;
    JPQLQuery<Event>query = from(event);
    switch (eventQueryParams.getType()) {
      case OWNER:
        if (user != null) {
          query = query.where(event.owner.eq(user));
        }
        break;
      case JOINER:
        if (user != null) {
          query = query.where(event.users.contains(user));
        }
        break;
      case ALL:
        break;
      default:
        throw new IllegalArgumentException("eventQueryParams.getType() can not be null");

    }

    // TODO Extract
    Optional<Coordinate> coordinate = eventQueryParams.getCoordinate();
    if (coordinate.isPresent()) {
      Coordinate exactCoordinate = coordinate.get();
      query = query.orderBy(
              (
                      (event.location.longitude.subtract(exactCoordinate.getLongitude()).abs())
                              .add(
                                      (event.location.latitude).subtract(exactCoordinate.getLatitude()).abs())
              )
                      .asc()
      );
    }
    query = query.where(conditionalStartedAt(pageable, eventQueryParams));


    final List<Event> events = query
            .offset(pageable.getOffset())
            .limit(pageable.getPageSize())
            .fetch();
    return new PageImpl<>(events, pageable, query.fetchCount());
  }

  @Override
  public Page<Event> search(Pageable pageable, EventQueryParams eventQueryParams) {
    return search(pageable, null, eventQueryParams);
  }

  private Predicate conditionalStartedAt(Pageable pageable, EventQueryParams params){
      if (params.getStartedAt() == null){
          return null;
      }
      return null;
  }
}

File: src/main/kotlin/social/alone/server/event/repository/EventRepositoryImpl.kt
package social.alone.server.event.repository

import com.querydsl.core.types.OrderSpecifier
import com.querydsl.core.types.dsl.BooleanExpression
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.support.QuerydslRepositorySupport
import social.alone.server.event.domain.Event
import social.alone.server.event.domain.QEvent.event
import social.alone.server.event.type.Coordinate
import social.alone.server.event.type.EventQueryParams
import java.time.LocalDateTime

class EventRepositoryImpl : QuerydslRepositorySupport(Event::class.java), EventRepositoryCustom {

    override fun search(pageable: Pageable, eventQueryParams: EventQueryParams): PageImpl<Event> {
        val query = from<Event>(event)
        // TODO Comment out, 개발시 지나간 모임도 보여야 편함
//        query.where(filterEndEvent())

        if (eventQueryParams.searchByCoordinate()){
            query.orderBy(sortIfCoordinate(eventQueryParams.coordinate!!))
        }

        this.querydsl!!.applyPagination(pageable, query)
        query.offset(pageable.offset)
        query.limit(pageable.pageSize.toLong())

        return PageImpl(query.fetch(), pageable, query.fetchCount())
    }

    private fun filterEndEvent(): BooleanExpression? {
        return event.startedAt.after(LocalDateTime.now())
    }

    private fun sortIfCoordinate(coordinate:Coordinate): OrderSpecifier<Double> {
        return (
                event.location.longitude.subtract(coordinate.longitude).abs()
                        .add(
                                event.location.latitude.subtract(coordinate.latitude).abs())
                        .asc()
                )
    }


}
